File: src/test/java/hm/binkley/labs/SimpleMagicBusTest.java
package hm.binkley.labs;

import hm.binkley.labs.MagicBus.FailedMessage;
import hm.binkley.labs.MagicBus.Mailbox;
import hm.binkley.labs.MagicBus.ReturnedMessage;
import lombok.NoArgsConstructor;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.*;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.function.Supplier;

import static hm.binkley.labs.SimpleMagicBus.ignored;
import static java.util.Collections.emptyList;
import static java.util.concurrent.TimeUnit.MILLISECONDS;
import static java.util.stream.IntStream.range;
import static lombok.AccessLevel.PRIVATE;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.awaitility.Awaitility.await;

@SuppressWarnings("rawtypes")
        // TODO: Clean up javac warns
class SimpleMagicBusTest {
    private final List<ReturnedMessage> returned
            = new CopyOnWriteArrayList<>();
    private final List<FailedMessage> failed = new CopyOnWriteArrayList<>();
    private final Map<Mailbox, List<Object>> observed = new LinkedHashMap<>();

    private MagicBus bus;

    private static <T> List<T> noMailbox() {
        return emptyList();
    }

    private static <T> Mailbox<T> record(final AtomicInteger order,
            final AtomicInteger record) {
        return __ -> record.set(order.getAndIncrement());
    }

    private static <T, E extends Exception> Mailbox<T> failWith(
            final Supplier<E> ctor) {
        return __ -> {
            throw ctor.get();
        };
    }

    @BeforeEach
    void setUp() {
        bus = SimpleMagicBus.of(returned::add, failed::add,
                (mailbox, message) -> observed
                        .computeIfAbsent(mailbox, __ -> new ArrayList<>())
                        .add(message));
    }

    @Test
    void shouldRejectMissingReturnedHandlerInConstructor() {
        assertThatThrownBy(
                () -> SimpleMagicBus.of(null, failed::add, ignored()))
                .isInstanceOf(NullPointerException.class);
    }

    @Test
    void shouldRejectMissingFailedHandlerInConstructor() {
        assertThatThrownBy(
                () -> SimpleMagicBus.of(returned::add, null, ignored()))
                .isInstanceOf(NullPointerException.class);
    }

    @Test
    void shouldRejectMissingObservedHandlerInConstructor() {
        assertThatThrownBy(
                () -> SimpleMagicBus.of(returned::add, failed::add, null))
                .isInstanceOf(NullPointerException.class);
    }

    @Test
    void shouldRejectMissingMessageTypeInSubscribe() {
        assertThatThrownBy(() -> bus.subscribe(null, __ -> {
        })).isInstanceOf(NullPointerException.class);
    }

    @Test
    void shouldRejectMissingMailboxInSubscribe() {
        assertThatThrownBy(() -> bus.subscribe(RightType.class, null))
                .isInstanceOf(NullPointerException.class);
    }

    @Test
    void shouldRejectMissingMessageTypeInUnsubscribe() {
        assertThatThrownBy(() -> bus.unsubscribe(null, __ -> {
        })).isInstanceOf(NullPointerException.class);
    }

    @Test
    void shouldRejectMissingMailboxInUnsubscribe() {
        assertThatThrownBy(() -> bus.unsubscribe(RightType.class, null))
                .isInstanceOf(NullPointerException.class);
    }

    @Test
    void shouldRejectMissingMessageInPublish() {
        assertThatThrownBy(() -> bus.post(null))
                .isInstanceOf(NullPointerException.class);
    }

    @Test
    void shouldReceiveCorrectType() {
        final var messages = new ArrayList<RightType>(1);
        bus.subscribe(RightType.class, messages::add);
        final var message = new RightType();
        bus.post(message);
        assertOn(messages).delivered(message).noneReturned().noneFailed();
    }

    @Test
    void shouldNotReceiveWrongType() {
        final var messages = new ArrayList<LeftType>(1);
        bus.subscribe(LeftType.class, messages::add);
        final var message = new RightType();
        bus.post(message);
        assertOn(messages).noneDelivered().returned(this.with(message))
                .noneFailed();
    }

    @Test
    void shouldReceiveSubtypes() {
        final var messages = new ArrayList<BaseType>(1);
        bus.subscribe(BaseType.class, messages::add);
        final var message = new RightType();
        bus.post(message);
        assertOn(messages).delivered(message).noneReturned().noneFailed();
    }

    @Test
    void shouldSaveDeadLetters() {
        final var message = new LeftType();
        bus.post(message);
        assertOn(noMailbox()).noneDelivered().returned(this.with(message))
                .noneFailed();
    }

    @Test
    void shouldSaveFailedPosts() {
        final var failure = new Exception();
        final Mailbox<LeftType> mailbox = failWith(() -> failure);
        bus.subscribe(LeftType.class, mailbox);
        final var message = new LeftType();
        bus.post(message);
        assertOn(noMailbox()).noneDelivered().noneReturned()
                .failed(this.with(mailbox, message, failure));
    }

    @Test
    void shouldThrowFailedPostsForUnchecked() {
        assertThatThrownBy(() -> {
            bus.subscribe(LeftType.class, failWith(RuntimeException::new));
            bus.post(new LeftType());
        }).isInstanceOf(RuntimeException.class);
    }

    @Test
    void shouldReceiveEarlierSubscribersFirst() {
        final var delivery = new AtomicInteger();
        final var second = new AtomicInteger();
        final var first = new AtomicInteger();
        final var fourth = new AtomicInteger();
        final var third = new AtomicInteger();
        bus.subscribe(RightType.class, record(delivery, first));
        bus.subscribe(RightType.class, record(delivery, second));
        bus.subscribe(RightType.class, record(delivery, third));
        bus.subscribe(RightType.class, record(delivery, fourth));
        bus.post(new RightType());
        assertThat(first.get()).isEqualTo(0);
        assertThat(second.get()).isEqualTo(1);
        assertThat(third.get()).isEqualTo(2);
        assertThat(fourth.get()).isEqualTo(3);
        assertOn(noMailbox()).noneReturned().noneFailed();
    }

    @Test
    void shouldReceiveOnParentTypeFirstInParentFirstOrder() {
        final var delivery = new AtomicInteger();
        final var farRight = new AtomicInteger();
        final var right = new AtomicInteger();
        final var base = new AtomicInteger();
        final var object = new AtomicInteger();
        bus.subscribe(RightType.class, record(delivery, right));
        bus.subscribe(FarRightType.class, record(delivery, farRight));
        bus.subscribe(Object.class, record(delivery, object));
        bus.subscribe(BaseType.class, record(delivery, base));
        bus.post(new FarRightType());
        assertThat(object.get()).isEqualTo(0);
        assertThat(base.get()).isEqualTo(1);
        assertThat(right.get()).isEqualTo(2);
        assertThat(farRight.get()).isEqualTo(3);
        assertOn(noMailbox()).noneReturned().noneFailed();
    }

    @Test
    void shouldUnsubscribeOnlyMailbox() {
        final var messages = new ArrayList<LeftType>(0);
        final Mailbox<LeftType> mailbox = messages::add;
        bus.subscribe(LeftType.class, mailbox);
        bus.unsubscribe(LeftType.class, mailbox);
        final var message = new LeftType();
        bus.post(message);
        assertOn(messages).noneDelivered().returned(this.with(message))
                .noneFailed();
    }

    @Test
    void shouldUnsubscribeRightMailbox() {
        final var messagesA = new ArrayList<RightType>(1);
        final Mailbox<RightType> mailboxB = failWith(Exception::new);
        bus.subscribe(RightType.class, messagesA::add);
        bus.subscribe(RightType.class, mailboxB);
        bus.unsubscribe(RightType.class, mailboxB);
        final var message = new RightType();
        bus.post(message);
        assertOn(messagesA).delivered(message).noneReturned().noneFailed();
    }

    @Test
    void shouldUnsubscribeThreadSafely() {
        await().atMost(2000L, MILLISECONDS).until(() -> {
            final var latch = new CountDownLatch(100);
            range(0, 100).parallel().forEach((actor) -> {
                final Mailbox<RightType> mailbox = new Discard();
                bus.subscribe(RightType.class, mailbox);
                bus.unsubscribe(RightType.class, mailbox);
                latch.countDown();
            });
            assertThat(latch.await(1000L, MILLISECONDS)).isNotEqualTo(0);
            final var message = new RightType();
            bus.post(message);
            assertOn(noMailbox()).returned(this.with(message)).noneFailed();
            return true;
        });
    }

    @Test
    void shouldComplainWhenUnsubscribingBadMailbox() {
        assertThatThrownBy(() -> {
            bus.subscribe(RightType.class, __ -> {
            });
            final Mailbox<RightType> mailbox = __ -> {
            };
            bus.unsubscribe(RightType.class, mailbox);
        }).isInstanceOf(NoSuchElementException.class);
    }

    @Test
    void shouldComplainWhenUnsubscribingBadMessageType() {
        assertThatThrownBy(() -> {
            final Mailbox<RightType> mailbox = __ -> {
            };
            bus.unsubscribe(RightType.class, mailbox);
        }).isInstanceOf(NoSuchElementException.class);
    }

    @Test
    void shouldProvideSubscriberForMessageType() {
        final Mailbox<RightType> a = new Mailbox<>() {
            @Override
            public void receive(final RightType message) {
            }

            @Override
            public String toString() {
                return "b";
            }
        };
        final Mailbox<BaseType> b = new Mailbox<>() {
            @Override
            public void receive(final BaseType message) {
            }

            @Override
            public String toString() {
                return "a";
            }
        };
        bus.subscribe(BaseType.class, b);
        bus.subscribe(RightType.class, a);

        assertThat(((SimpleMagicBus) bus).subscribers(RightType.class))
                .isEqualTo(List.of(b, a));
    }

    private <T> AssertDelivery<T> assertOn(final List<T> delivered) {
        return new AssertDelivery<>(delivered);
    }

    private ReturnedMessage with(final Object message) {
        return new ReturnedMessage(bus, message);
    }

    private FailedMessage with(final Mailbox<?> mailbox, final Object message,
            final Exception failure) {
        return new FailedMessage(bus, mailbox, message, failure);
    }

    private static final class FarRightType
            extends RightType {
    }

    private static class RightType
            extends BaseType {
    }

    private static final class LeftType
            extends BaseType {
    }

    private abstract static class BaseType {
        BaseType() {
        }
    }

    @NoArgsConstructor(access = PRIVATE)
    private static class Discard
            implements Mailbox<RightType> {
        public void receive(final RightType message) {
        }
    }

    private final class AssertDelivery<T> {
        private final List<T> delivered;

        // TODO: Javac is choking when using @RequiredArgsCtor
        @java.beans.ConstructorProperties({"delivered"})
        private AssertDelivery(final List<T> delivered) {
            this.delivered = delivered;
        }

        private AssertDelivery<T> noneDelivered() {
            assertThat(this.delivered).isEmpty();
            return this;
        }

        @SafeVarargs
        @SuppressWarnings("varargs")
        private <U extends T> AssertDelivery<T> delivered(
                final U... delivered) {
            assertThat(this.delivered).containsExactly(delivered);
            return this;
        }

        private AssertDelivery<T> noneReturned() {
            assertThat(SimpleMagicBusTest.this.returned).isEmpty();
            return this;
        }

        private AssertDelivery<T> returned(
                final ReturnedMessage... returned) {
            assertThat(SimpleMagicBusTest.this.returned)
                    .containsExactly(returned);
            return this;
        }

        private AssertDelivery<T> noneFailed() {
            assertThat(SimpleMagicBusTest.this.failed).isEmpty();
            return this;
        }

        private AssertDelivery<T> failed(final FailedMessage... failed) {
            assertThat(SimpleMagicBusTest.this.failed)
                    .containsExactly(failed);
            return this;
        }
    }
}

File: src/test/kotlin/hm/binkley/labs/SimpleMagicBusTest.kt
package hm.binkley.labs

import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import java.nio.charset.CoderMalfunctionError
import java.util.NoSuchElementException
import java.util.concurrent.atomic.AtomicInteger

internal class SimpleMagicBusTest {
    private val bus = TestMagicBus()

    @Test
    fun `should have distinct named mailboxen even with the same name`() {
        assertThat(namedMailbox<RightType>("BOB") {})
            .isNotEqualTo(namedMailbox<RightType>("BOB") {})
    }

    @Test
    fun `should do nothing when unsubscribing`() {
        val mailbox = testMailbox<RightType>()
            .subscribeTo(bus)
            .unsubscribeFrom(bus)
        val message = RightType()

        message postTo bus

        assertOn(mailbox)
            .noDeliveredMessages()
            .undeliveredInOrder(message)
            .noFailedMessages()
    }

    @Test
    fun `should stop tracking message types with no subscribers`() {
        testMailbox<RightType>()
            .subscribeTo(bus)
            .unsubscribeFrom(bus)

        assertThat(bus.subscriptions.containsKey(RightType::class.java))
            .isFalse
    }

    @Test
    fun `should receive correct type`() {
        val mailbox = testMailbox<RightType>() subscribeTo bus
        val message = RightType()

        message postTo bus

        assertOn(mailbox)
            .deliveredInOrder(message)
            .noUndeliveredMessages()
            .noFailedMessages()
    }

    @Test
    fun `should deliver correctly to disparate subscribers`() {
        val mailboxForDelivery = testMailbox<RightType>() subscribeTo bus
        val mailboxForUndelivered = testMailbox<LeftType>() subscribeTo bus
        val mailboxForAliens = testMailbox<AlienType>() subscribeTo bus

        assertThat(bus.subscribersTo<RightType>())
            .containsOnly(mailboxForDelivery)

        val message = RightType()
        val alienMessage = AlienType()

        message postTo bus
        alienMessage postTo bus

        assertOn(mailboxForDelivery)
            .deliveredInOrder(message)
            .noUndeliveredMessages()
            .noFailedMessages()
        assertOn(mailboxForUndelivered)
            .noDeliveredMessages()
            .noUndeliveredMessages()
            .noFailedMessages()
        assertOn(mailboxForAliens)
            .deliveredInOrder(alienMessage)
            .noUndeliveredMessages()
            .noFailedMessages()
    }

    @Test
    fun `should not receive wrong type`() {
        val mailbox = testMailbox<LeftType>() subscribeTo bus
        val message = RightType()

        message postTo bus

        assertOn(mailbox)
            .noDeliveredMessages()
            .undeliveredInOrder(message)
            .noFailedMessages()
    }

    @Test
    fun `should receive subtypes`() {
        val mailbox = testMailbox<BaseType>() subscribeTo bus
        val message = RightType()

        message postTo bus

        assertOn(mailbox)
            .deliveredInOrder(message)
            .noUndeliveredMessages()
            .noFailedMessages()
    }

    @Test
    fun `should save dead letters`() {
        val message = LeftType()

        message postTo bus

        assertOn(allMailboxen())
            .noDeliveredMessages()
            .undeliveredInOrder(message)
            .noFailedMessages()
    }

    @Test
    fun `should provide distinct discard mailboxen`() {
        assertThat(discard<RightType>()).isNotEqualTo(discard<RightType>())
    }

    @Test
    fun `should provide distinct failure mailboxen`() {
        assertThat(failWith<RightType> { Exception() })
            .isNotEqualTo(failWith<RightType> { Exception() })
    }

    @Test
    fun `should post failed messages separately for each mailbox`() {
        val failureA = Exception("A")
        val brokenMailboxA = failWith<LeftType> { failureA } subscribeTo bus
        val failureB = Exception("B")
        val brokenMailboxB = failWith<LeftType> { failureB } subscribeTo bus
        val message = LeftType()

        message postTo bus

        assertOn(allMailboxen())
            .noDeliveredMessages()
            .noUndeliveredMessages()
            .failedInOrder(
                brokenMailboxA with message and failureA,
                brokenMailboxB with message and failureB,
            )
    }

    @Test
    fun `should bubble out runtime exceptions`() {
        val failure = RuntimeException()
        bus += failWith<LeftType> { failure }

        assertThatThrownBy {
            LeftType() postTo bus
        }.isSameAs(failure)
    }

    @Test
    fun `should let JVM errors (Error vs Exception) bubble out`() {
        val failure = CoderMalfunctionError(
            Exception("SOMETHING WICKED THIS WAY COMES")
        )
        bus += failWith<LeftType> { failure }

        assertThatThrownBy {
            LeftType() postTo bus
        }.isSameAs(failure)
    }

    @Test
    fun `should post other exceptions and receive them`() {
        val failure = Exception()
        val badMailbox = failWith<RightType> { failure } subscribeTo bus
        val allMailbox = testMailbox<Any>() subscribeTo bus

        val message = RightType()
        val failedMessage = FailedMessage(bus, badMailbox, message, failure)
        val returnReceipt = ReturnReceipt(bus, message)

        message postTo bus

        assertOn(allMailbox)
            .noUndeliveredMessages()
            .deliveredInOrder(message, failedMessage, returnReceipt)
            .failedInOrder(badMailbox with message and failure)
    }

    @Test
    fun `should be fatal for a failure mailbox to itself fail`() {
        bus += failWith<RightType> { Exception() }
        bus += failWith<FailedMessage<Any>> { Exception() }

        // TODO: Rather than overflowing the stack, how to alert caller?
        //       Not appropriate to add a logging framework, so what else?
        assertThrows<StackOverflowError> {
            RightType() postTo bus
        }
    }

    @Test
    fun `should be fatal for a mailbox to repost in an infinite cycle`() {
        bus += namedMailbox<RightType>("REPOSTING-RIGHT") {
            it postTo bus
        }

        // TODO: Rather than overflowing the stack, how to alert caller?
        //       Not appropriate to add a logging framework, so what else?
        assertThrows<StackOverflowError> {
            RightType() postTo bus
        }
    }

    @Test
    fun `should receive mailboxen for same type in subscription order`() {
        val mailboxen = OrderedMailboxen()
        val mailboxA = mailboxen.orderedMailbox<RightType>() subscribeTo bus
        val mailboxB = mailboxen.orderedMailbox<RightType>() subscribeTo bus

        RightType() postTo bus

        assertThat(mailboxen.deliveriesInOrder()).isEqualTo(
            listOf(
                mailboxA,
                mailboxB,
            )
        )
    }

    @Test
    fun `should receive parent types before child types`() {
        // FYI -- it is important that mailboxen subscribe *not* in class
        // hierarchy order, so that the test can verify messages are received
        // in the correct order nonetheless
        val mailboxen = OrderedMailboxen()
        val rightMailbox =
            mailboxen.orderedMailbox<RightType>() subscribeTo bus
        //  Add an unrelated type which should *not* show up
        mailboxen.orderedMailbox<LeftType>() subscribeTo bus
        val farRightMailbox =
            mailboxen.orderedMailbox<FarRightType>() subscribeTo bus
        val receiptsMailbox =
            mailboxen.orderedMailbox<ReturnReceipt<*>>() subscribeTo bus
        val baseMailbox =
            mailboxen.orderedMailbox<BaseType>() subscribeTo bus
        val allMailbox =
            mailboxen.orderedMailbox<Any>() subscribeTo bus
        //  And another unrelated type which should *not* show up
        mailboxen.orderedMailbox<LeftType>() subscribeTo bus

        FarRightType() postTo bus

        assertThat(mailboxen.deliveriesInOrder()).isEqualTo(listOf(
            allMailbox,
            baseMailbox,
            rightMailbox,
            farRightMailbox,
            receiptsMailbox,
        ))
    }

    @Test
    fun `should unsubscribe exact type`() {
        val mailbox = discard<LeftType>() subscribeTo bus

        mailbox unsubscribeFrom bus

        assertThat(bus.subscribersTo<LeftType>()).isEmpty()
    }

    @Test
    fun `should unsubscribe exact mailbox`() {
        val mailboxForDelivery = discard<RightType>() subscribeTo bus
        val mailboxForUndelivered = discard<RightType>() subscribeTo bus

        bus -= mailboxForUndelivered

        assertThat(bus.subscribersTo<RightType>())
            .containsOnly(mailboxForDelivery)
    }

    @Test
    fun `should fail to unsubscribe for message type`() {
        val mailbox = discard<RightType>()

        assertThrows<NoSuchElementException> {
            bus -= mailbox
        }
    }

    @Test
    fun `should fail to unsubscribe exact mailbox regardless of other mailboxen`() {
        bus += discard<RightType>()
        val mailboxNotSubscribed = discard<RightType>()

        assertThrows<NoSuchElementException> {
            bus -= mailboxNotSubscribed
        }
    }

    @Test
    fun `should provide subscribers for message type in base-up order`() {
        // Subscription in derived-base order to show that the subscriber
        // list is in base-derived order
        val mailboxRight = testMailbox<RightType>() subscribeTo bus
        val mailboxBase = testMailbox<BaseType>() subscribeTo bus

        assertThat(bus.subscribersTo<RightType>())
            .containsOnly(mailboxBase, mailboxRight)
    }

    @Test
    fun `should provide subscribers for message type in subscribed-order`() {
        val mailboxen = (1..10).map {
            // An interesting example of when the return type is needed:
            // 1) Kotlin returns the final value of the lambda
            // 2) `bus += mailbox` return Unit
            // 3) `subscribeTo` returns the mailbox, fluent style
            // 4) The test wants to compare mailboxen, so Unit is wrong
            testMailbox<RightType>() subscribeTo bus
        }

        // It is unusual to test the test code, but the production test relies
        // critically on test mailbox instances being non-equal
        assertThat(testMailbox<RightType>())
            .isNotEqualTo(testMailbox<RightType>())
        assertThat(bus.subscribersTo<RightType>()).isEqualTo(mailboxen)
    }

    @Test
    fun `should provide accurate details on dead letters`() {
        val message = RightType()

        message postTo bus

        assertThat(bus.undelivered).containsOnly(
            UndeliveredMessage(
                bus,
                message
            )
        )
    }

    @Test
    fun `should have default rejected letter box`() {
        val defaultMailboxCountForUndeliveredAndFailedMessages = 2

        assertThat(bus.subscribersTo<FailedMessage<Any>>())
            .hasSize(defaultMailboxCountForUndeliveredAndFailedMessages)
    }

    @Test
    fun `should deliver to first subscriber for undelivered messages before default`() {
        val subscribers = bus.subscribersTo<UndeliveredMessage<Any>>()

        assertThat(subscribers.first().toString())
            .isEqualTo("DISCARD-MAILBOX<UndeliveredMessage>")
    }

    @Test
    fun `should deliver to first subscriber for failed messages before default`() {
        val subscribers = bus.subscribersTo<FailedMessage<Any>>()

        assertThat(subscribers.first().toString())
            .isEqualTo("DISCARD-MAILBOX<FailedMessage>")
    }

    @Test
    fun `should discard in the discard letter box`() {
        bus += discard<RightType>()

        RightType() postTo bus

        assertOn(allMailboxen())
            .noDeliveredMessages()
            .noUndeliveredMessages()
            .noFailedMessages()
    }

    @Test
    fun `should provide accurate details on failed posts`() {
        val failure = Exception()
        val mailbox = failWith<RightType> { failure } subscribeTo bus
        val message = RightType()

        message postTo bus

        assertThat(bus.failed)
            .containsOnly(FailedMessage(bus, mailbox, message, failure))
    }

    @Test
    fun `should have named mailboxen`() {
        val name = "BOB'S YER UNKEL"
        val mailbox = namedMailbox<RightType>(name) {}

        assertThat(mailbox.toString()).isEqualTo(name)
    }

    private fun <T> assertOn(vararg delivered: TestMailbox<T>) =
        assertOn(delivered.asList())

    private fun <T> assertOn(delivered: List<TestMailbox<T>>) =
        AssertDelivered(
            delivered.flatMap { it.messages },
            bus.undelivered,
            bus.failed
        )

    private infix fun <T> Mailbox<T>.with(message: T) = this to message
    private infix fun <T> Pair<Mailbox<T>, T>.and(failure: Exception) =
        this to failure

    private inline fun <reified T> testMailbox(
        messages: MutableList<T> = mutableListOf(),
    ) = TestMailbox(bus, T::class.java, messages)

    private fun allMailboxen() = listOf<TestMailbox<Any>>()

    private inner class AssertDelivered<T>(
        private val delivered: List<T>,
        private val undelivered: List<UndeliveredMessage<*>>,
        private val failed: List<FailedMessage<*>>,
    ) {
        fun noDeliveredMessages() = apply {
            assertThat(delivered).isEmpty()
        }

        fun <U : T> deliveredInOrder(vararg delivered: U) = apply {
            assertThat(this.delivered).isEqualTo(delivered.toList())
        }

        fun noUndeliveredMessages() = apply {
            assertThat(undelivered).isEmpty()
        }

        fun undeliveredInOrder(vararg undelivered: Any) = apply {
            assertThat(this.undelivered).isEqualTo(
                undelivered.map {
                    UndeliveredMessage(bus, it)
                }
            )
        }

        fun noFailedMessages() = apply {
            assertThat(failed).isEmpty()
        }

        fun <U : T> failedInOrder(
            vararg failed: Pair<Pair<Mailbox<U>, U>, Exception>,
        ) = apply {
            assertThat(this.failed).isEqualTo(
                failed.map {
                    val (first, failure) = it
                    val (mailbox, message) = first
                    FailedMessage(bus, mailbox, message, failure)
                }
            )
        }
    }
}

private class OrderedMailboxen {
    private val sequence = AtomicInteger(-1)
    private val mailboxen = mutableListOf<OrderedMailbox<*>>()

    inline fun <reified T> orderedMailbox(): OrderedMailbox<T> =
        OrderedMailbox(T::class.java).also { mailboxen.add(it) }

    fun deliveriesInOrder(): List<Mailbox<*>> = mailboxen
        .filter { -1 < it.order } // -1 means no deliveries
        .sortedBy {
            it.order
        }.also {
            println("Mailboxen order: $it")
        }

    inner class OrderedMailbox<T>(
        private val messageType: Class<T>,
    ) : Mailbox<T> {
        private val myOrder: AtomicInteger = AtomicInteger(-1)
        val order: Int get() = myOrder.get()

        override fun invoke(message: T) =
            myOrder.set(sequence.incrementAndGet())

        override fun toString() =
            "TEST-ORDERED-MAILBOX<${messageType.simpleName}@$myOrder>"
    }
}

private fun <T : Any> failWith(failure: () -> Throwable): Mailbox<T> =
    { throw failure() }

private abstract class BaseType
private class LeftType : BaseType()
private open class RightType : BaseType()
private class FarRightType : RightType()
private class AlienType

/**
 * @todo Why is delivered key `<*>` but the value is `<Any>`?
 * @todo Why is undelivered `<Any>` but failed is `<*>`?
 */
private class TestMagicBus : SimpleMagicBus() {
    val delivered = mutableMapOf<Mailbox<*>, MutableList<Any>>()
    private val _undelivered = mutableListOf<UndeliveredMessage<Any>>()
    private val _failed = mutableListOf<FailedMessage<*>>()

    val undelivered: List<UndeliveredMessage<Any>> get() = _undelivered
    val failed: List<FailedMessage<*>> get() = _failed

    init {
        this += namedMailbox<UndeliveredMessage<Any>>("TEST-DEAD-LETTERBOX") {
            _undelivered += it
        }
        this += namedMailbox<FailedMessage<Any>>("TEST-FAILED-LETTERBOX") {
            _failed += it
        }
    }
}

private class TestMailbox<T>(
    private val bus: TestMagicBus,
    private val messageType: Class<T>,
    val messages: MutableList<T>,
) : Mailbox<T> {
    override operator fun invoke(message: T) {
        messages.add(message)
        bus.delivered.getOrPut(this) {
            mutableListOf()
        } += message as Any
    }

    override fun toString() = "TEST-MAILBOX<${messageType.simpleName}>"
}
